File: scripts/wordclouds.R
library(dplyr)
library(tidytext)
library(tidyr)
library(tm)
library(wordcloud)

source("load_texts.R")

numberOfWords <- 20

df <-return_data_frame()

wordsFreq <- df %>%
    unnest_tokens(word, text) %>%
    count(word, sort = TRUE) %>%
    ungroup()

wordsFreq <- wordsFreq %>%
    filter(!word %in% Mystopwords)
print(wordsFreq[1:numberOfWords, ])

wordcloud(
    words = wordsFreq$word, freq = wordsFreq$n, min.freq = 1000,
    random.order = FALSE, max.words = 1000, rot.per = 0,
    colors = brewer.pal(6, "Dark2"), scale = c(4, 0.8)
)

Mystopwords <- c(Mystopwords, "inflação", "preços", "taxa", "copom", "bilhões", "crescimento",
                 "bens","produção", "cenário", "política", "monetária", "período", "anterior",
                 "us","p.p","índice","mercado","comitê","atividade", "acordo", "mensal", "leva",
                 "janeiro","fevereiro","março","abril","maio","junho","julho","agosto",
                 "setembro","outubro","novembro","dezembro","econômica","economia","mês",
                 "taxas", "vista", "_","-","reunião","bem","relação","primeiro","segundo",
                 "terceiro","quarto","luiz","tal","dia","paulo","diz","vem")

df2 <- df[,c("meeting","text","decision")]
colnames(df2) <- c("doc_id","text","decision")
lower <- df2[df2$decision == "lower",]
raise <- df2[df2$decision == "raise",]
keep <- df2[df2$decision == "keep",]

lower <- as.vector(lower$text)
raise <- as.vector(raise$text)
keep <- as.vector(keep$text)

All <- list(lower, raise, keep)

docs <- Corpus(VectorSource(All))
toSpace <- content_transformer(function(x, pattern) gsub(pattern, " ", x))
docs <- tm_map(docs, content_transformer(tolower))
docs <- tm_map(docs, removeNumbers)
docs <- tm_map(docs, removePunctuation)
docs <- tm_map(docs, removeWords, Mystopwords)
docs <- tm_map(docs, stripWhitespace)

tdm <- TermDocumentMatrix(docs)
tdm = removeSparseTerms(tdm, 0.98)
tdm = as.matrix(tdm)
colnames(tdm) <- c("redução", "aumento", "manutenção")

comparison.cloud(tdm, max.words = 500, random.order = FALSE, rot.per = 0, scale = c(2, 0.5), fixed.asp=T, title.size = 1)
commonality.cloud(tdm, max.words = 300, random.order = FALSE, rot.per = 0, scale = c(2, 0.5))

File: scripts/wordclouds.py
import re
import wordcloud
import matplotlib.pyplot as plt
from load_texts import *  # return_data_frame() and Mystopwords

Mystopwords = Mystopwords + ["anterior", "banco", "central", "copom", "doze", "índice", "inflação", "meta", "monetária", "período", "política", 
    "preço", "preços", "relação", "taxa"]

def wordcloudPlot(text, stopwords=None, title = "", max_font_size=50, max_words=100, background_color="white"):
    cloud = wordcloud.WordCloud(stopwords=stopwords, max_font_size=max_font_size,
                                max_words=max_words, background_color=background_color).generate(text.lower())

    # Display the generated image:
    plt.imshow(cloud, interpolation='bilinear')
    plt.title(title)
    plt.axis("off")
    plt.show()


print("Loading data...")
dfCorpus = return_data_frame()
dfCorpus.text = dfCorpus.text.apply(lambda x : x.lower())
corpus = dfCorpus.text.to_list()
print(len(corpus), "minutes")

corpusLower = dfCorpus[dfCorpus.decision == "lower"].text.to_list()
print(len(corpusLower), "minutes - lower")
corpusRaise = dfCorpus[dfCorpus.decision == "raise"].text.to_list()
print(len(corpusRaise), "minutes - raise")
corpusKeep = dfCorpus[dfCorpus.decision == "keep"].text.to_list()
print(len(corpusKeep), "minutes - keep")

corpusJoined = ' '.join(corpus)
corpusJoined = corpusJoined.lower()
corpusJoined = re.sub('\n', '', corpusJoined)  

corpusLowerJoined = ' '.join(corpusLower)
corpusLowerJoined = corpusLowerJoined.lower()
corpusLowerJoined = re.sub('\n', '', corpusLowerJoined)  

corpusRaiseJoined = ' '.join(corpusRaise)
corpusRaiseJoined = corpusRaiseJoined.lower()
corpusRaiseJoined = re.sub('\n', '', corpusRaiseJoined)  

corpusKeepJoined = ' '.join(corpusKeep)
corpusKeepJoined = corpusKeepJoined.lower()
corpusKeepJoined = re.sub('\n', '', corpusKeepJoined)  

wordcloudPlot(corpusJoined, stopwords=Mystopwords, title="Minutes")
wordcloudPlot(corpusLowerJoined, stopwords=Mystopwords, title="Lower")
wordcloudPlot(corpusRaiseJoined, stopwords=Mystopwords, title="Raise")
wordcloudPlot(corpusKeepJoined, stopwords=Mystopwords, title="Keep")
